Refactors chicks into basic mobs

san7890 opened the pull request:
## About The Pull Request

On the tin. They have pretty much nothing in common with chickens, so no subtyping. They are in the same folder to keep that whole thing tidy, though.

Also includes fixes to `growth_and_differentiation` element that I made for spiderlings, since some stuff was yorked without me realizing. It pretty much worked flawlessly for these chicks otherwise though. It all works fine now.
## Why It's Good For The Game

More verbose naming scheme (instead of "holo", we get "permanent" chicks), smarter AI for chicks, knocks them off the list, etc. etc.

One thing that I wanted to do was to have chicks recognize their mother (if they had one), but that would be way out of scope for this simple port PR. I'll dwell on adding something cool for that in the future.
## Changelog
:cl:
refactor: Chicks are now a bit smarter, be careful not to squish them!
/:cl:

Let me know if the whole "COMPONENT_KILL" thing is cringe, I couldn't figure out a better way to do it without abusing `GetComponent()` to `qdel()` it that way.

san7890 committed:
first implementation

san7890 committed:
cleanup

san7890 committed:
find and replace

san7890 committed:
updatepaths

san7890 committed:
map updates

i decided to err on the side of caution and make it a permanent chick, because it has a name. it would lose its name every single round otherwise (and was losing its name every single round), but let's just be kind.

san7890 committed:
whoops wrong variable here

san7890 committed:
fixes some faulty assertions

these two were always broken i'm glad i figured it out now

san7890 committed:
Merge remote-tracking branch 'upstream/master' into hooters

san7890 committed:
Merge remote-tracking branch 'upstream/master' into hooters

san7890 committed:
yowzers

LemonInTheDark commented on code/__DEFINES/dcs/signals/signals_datum.dm:
```diff
@@ -7,6 +7,9 @@
 /// before a component is removed from a datum because of ClearFromParent: (/datum/component)
 #define COMSIG_COMPONENT_REMOVING "component_removing"
+/// when a datum wants to kill its component for any reason: (/datum/component)
+#define COMSIG_COMPONENT_KILL "component_kill"
```
yeah this is cringe. this is a proc, not a signal.

san7890 committed:
updates updatepaths

LemonInTheDark commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -44,6 +41,8 @@
 	src.optional_checks = optional_checks
 	src.optional_grow_behavior = optional_grow_behavior
 
+	RegisterSignal(parent, COMSIG_COMPONENT_KILL, PROC_REF(kill_processes))
```
if you don't want to use get component, allow users of the component to pass in signals that trigger the component disabling/deleting itself

LemonInTheDark commented on code/modules/mob/living/basic/farm_animals/chicken/chick.dm:
```diff
@@ -0,0 +1,74 @@
+
+/mob/living/basic/chick/Login()
+	. = ..()
+	SEND_SIGNAL(src, COMSIG_COMPONENT_KILL)
```
then just use the login signal rather then this proc call

san7890 committed:
switches it to passing in a list of signals

san7890 committed:
kill ambiguity

san7890 committed:
spacing nit

LemonInTheDark commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -41,7 +52,10 @@
+	if(islist(signals_to_kill_on))
+		src.signals_to_kill_on = signals_to_kill_on
+		for(var/signal in signals_to_kill_on)
+			RegisterSignal(parent, signal, PROC_REF(kill_processes))
```
may I recommend `RegisterSignals()`

san7890 replied:
brain fart moment

LemonInTheDark commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -56,6 +70,8 @@
 
 /datum/component/growth_and_differentiation/UnregisterFromParent()
 	UnregisterSignal(SSticker, COMSIG_TICKER_ROUND_STARTING)
+	for(var/signal in signals_to_kill_on)
+		UnregisterSignal(parent, signal)
```
why do we implement this in the first place? will this component be moved around at any point?

LemonInTheDark replied:
Also you can just pass a list into unreg signal and it'll work, it's even slightly faster

san7890 replied:
no because we qdel ourselves when growth is complete. i think admin schenanigans are the only real reason.

LemonInTheDark replied:
you don't need to support admin component transferring if you don't want to, I'm not sure we do in the first place

san7890 replied:
it doesn't seem like the hardest thing to support

MrMelbert commented on code/datums/ai/basic_mobs/basic_subtrees/speech_subtree.dm:
```diff
@@ -95,6 +95,11 @@
 	speak = list("Cluck!", "BWAAAAARK BWAK BWAK BWAK!", "Bwaak bwak.")
 	emote_hear = list("clucks.", "croons.")
 	emote_see = list("pecks at the ground.","flaps her wings viciously.")
+/datum/ai_planning_subtree/random_speech/chick
```
```suggestion
	emote_see = list("pecks at the ground.","flaps her wings viciously.")
	
/datum/ai_planning_subtree/random_speech/chick
```

LemonInTheDark replied:
newline?

MrMelbert commented on code/modules/cargo/packs/livestock.dm:
```diff
@@ -52,7 +52,7 @@
 	cost = CARGO_CRATE_VALUE * 4
 	access_view = ACCESS_KITCHEN
-	contains = list( /mob/living/simple_animal/chick)
+	contains = list( /mob/living/basic/chick)
```
```suggestion
	contains = list(/mob/living/basic/chick)
```

san7890 replied:
yeah this was done on an unpushed commit sorry

MrMelbert commented on code/modules/mob/living/basic/farm_animals/chicken/chick.dm:
```diff
@@ -0,0 +1,67 @@
+
+/// We don't grow into a chicken if we're not conscious or if we have a client (in case of schenanigans).
+/mob/living/basic/chick/proc/ready_to_grow()
+	if(stat == CONSCIOUS)
+		return TRUE
+
+	return FALSE
```
```suggestion
	return (stat == CONSCIOUS)
```

MrMelbert replied:
I also just noticed the documenting coment states `or if we have a client` - 
If it's outdated the comment should be updated, and if it's not updated you should probably add that code in.

san7890 replied:
it's not outdated it's just done elsewhere. i'll move that bit.

san7890 committed:
oh right they accept lists

MrMelbert commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -52,11 +65,19 @@
+
+/// In case the mob decides that we shouldn't grow anymore (permanently), we don't really wanna waste up the tick doing useless work. Let's stop everything and qdel ourselves.
+/datum/component/growth_and_differentiation/proc/kill_processes()
+	UnregisterFromParent()
+	STOP_PROCESSING(SSdcs, src)
+	qdel(src)
```
`UnregisterFromParent` is called from `Destroy` anyways
Likewise `Destroy` should have the code to stop processing

LemonInTheDark commented on code/modules/mob/living/basic/farm_animals/chicken/chick.dm:
```diff
@@ -0,0 +1,67 @@
+	gold_core_spawnable = FRIENDLY_SPAWN
+
+	/// What we grow into.
+	var/grow_as = /mob/living/basic/chicken
```
menacing

MrMelbert commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -44,6 +52,11 @@
 	src.optional_checks = optional_checks
 	src.optional_grow_behavior = optional_grow_behavior
 
+	if(islist(signals_to_kill_on))
+		src.signals_to_kill_on = signals_to_kill_on
+		for(var/signal in src.signals_to_kill_on)
+			RegisterSignal(parent, signal, PROC_REF(kill_processes))
+
```
Should go into `RegisterWithParent`? Doesn't matter since the comp is not transferable but you're using `UnregisterFromParent` so, consistency

MrMelbert commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -32,7 +31,16 @@
 
-/datum/component/growth_and_differentiation/Initialize(growth_time, growth_path, growth_probability, lower_growth_value, upper_growth_value, optional_checks, optional_grow_behavior)
+/datum/component/growth_and_differentiation/Initialize(
+	growth_time,
+	growth_path,
+	growth_probability,
+	lower_growth_value,
+	upper_growth_value,
+	signals_to_kill_on,
+	optional_checks,
+	optional_grow_behavior,
```
```suggestion
	growth_time,
	growth_path,
	growth_probability,
	lower_growth_value,
	upper_growth_value,
	list/signals_to_kill_on,
	datum/callback/optional_checks,
	datum/callback/optional_grow_behavior,
```

san7890 committed:
newline

MrMelbert commented:
Oof I got sniped

san7890 committed:
comment shuffle and code cleanup

san7890 committed:
types Initialize() args

san7890 committed:
moves everything to Destroy, invokes qdel instead

LemonInTheDark commented on code/datums/components/growth_and_differentiation.dm:
```diff
@@ -44,6 +52,10 @@
 
+	if(islist(signals_to_kill_on))
+		src.signals_to_kill_on = signals_to_kill_on
+		RegisterSignals(parent, src.signals_to_kill_on.Copy(), CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(qdel), src))
```
uh, what?

san7890 replied:
what's wrong specifically

san7890 replied:
oh it doesn't work at all one second

san7890 replied:
for some reason i thought i was doing timer stuff

san7890 committed:
makes the path-check more explicit + fixes the qdel

whoops got a bit too evil

san7890 committed:
wrapper proc instead

AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

san7890 committed:
fix a silly comment

JohnFulpWillard commented on code/modules/holodeck/holo_effect.dm:
```diff
@@ -83,8 +83,8 @@
 	. = ..()
 	mobtype = list(
 		/mob/living/basic/butterfly,
+		/mob/living/basic/chick/permanent,
```
why is a holodeck mob permanent?
